gatsby/create-node/create-preview-cards.js

```diff
@@ -1,65 +1,51 @@
-const slugify = require('slugify');
-const path = require('path');
-const { siteMetadata } = require('../../gatsby-config');
 const {
-  generateCard,
+  generateCardToBuffer,
 } = require('../util/preview-card-generator/generate-card');
 
+const { createFileNodeFromBuffer } = require('gatsby-source-filesystem');
+
 /**
  * Create a preview card file and put the url to the Gatsby store to be able
  * to link it in the page header.
  */
-const createPreviewMarkdown = ({ node, _, actions }) => {
-  const { createNodeField } = actions;
-  const post = node.frontmatter;
+const createPreviewCard = async (
+  title,
+  { node, _, actions, getCache, createNodeId },
+) => {
+  const { createNode, createNodeField } = actions;
 
   // we need the title in order to generate anything
-  if (!post.title) {
+  if (!title) {
     return;
   }
 
-  // we create a file name from the path (which is the page slug and more unique) but if none is available take the title
-  const fileName = `social-card---${slugify(post.path || post.title, {
-    lower: true,
-  })}.jpg`;
-  const outputFile = path.join('public', fileName);
-  const imagePath = `/${fileName}`;
+  const buffer = await generateCardToBuffer({ title });
 
-  generateCard({ title: post.title }, outputFile).then(() => {
-    createNodeField({
-      node,
-      name: `socialCard`,
-      value: imagePath,
-    });
+  const fileNode = await createFileNodeFromBuffer({
```

Comment:
This function creates AND links the node to the parent, right? A comment above would make this more clear, as I was just wondering at which place the actual connection between the Markdown/SyJob and the file is happening.

Reply:
I'll  add the following comment to make things clear:

> The util function `createFileNodeFromBuffer` from the official gatsby source plugin `gatsby-source-filesystem`
> creates a file node from a given file buffer. The value of `parentNodeId` creates the necessary relationship
> between the original node and the actual file node so it's not garbage collected.
>  
> The actual foreign key relationship is resolved through `createSchemaCustomization`
> in gatsby-node.js for all node types the `createPreviewCard` is invoked for.


Ideally  we create a plugin "create social cards" or such that would bundle this invocation (tied to `onCreateNodde`) and the fk lookup which is tied to `createSchemaCustomization`. Feels like a random split while it's technically required. That's our next small local plugin we can extract (and configure) 🙏